### ISSUE TITLE
`unnecessary_concatenation_linter` wrongly complains about usage with magrittr's assignment pipe operator `%<>%`

### ISSUE THREAD

**Issue by salim-b** [r-lib/lintr#2008]
Consider the following reprex:

``` r
tmp_file <- tempfile(fileext = "R")
writeLines(con = tmp_file,
           text = "
library(magrittr)
x <- 1:2
x %<>% c(3)
x
")

lintr::lint(filename = tmp_file,
            linters = list(lintr::unnecessary_concatenation_linter()))
#> /tmp/RtmpyaBMnL/filee0674c13fd78R:4:8: style: [unnecessary_concatenation_linter] Unneeded concatenation of a constant. Remove the "c" call.
#> x %<>% c(3)
#>        ^~~~
```

<sup>Created on 2023-07-20 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

`x %<>% c(3)` above is the [magrittr way](https://magrittr.tidyverse.org/reference/compound.html) to write `x <- c(x, 3)`, so lintr should ideally allow `c()` with a single constant *if* it is preceded by the `%<>%` operator.

**Comment by MichaelChirico** [comment]
Thanks Salim. I'm not exactly sure how to proceed.

On the one hand, the assignment pipe is not allowed by the Tidyverse guide:

https://style.tidyverse.org/pipes.html#assignment-2

on the other hand, that's kind of independent of the concatenation linter.

Probably we should:

 1. lint this with assignment_linter()
 2. not lint it with unnecessary_concatenation_linter()

**Comment by salim-b** [comment]
I know the TSG instructs to avoid this operator, but I still ike to use it. Thus I very much agree with your second action point.
